flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+                                                                                            bufferIndex));
+                                                }
+                                                subpartitionDataBuffer.onBufferReleased(
+                                                        bufferIndex);
+                                            });
+                                }
+                            });
+                });
+    }
+
+    private void recycleBuffer(MemorySegment buffer) {
+        numRequestedBuffers.decrementAndGet();
+        bufferPool.recycle(buffer);
+    }
+
+    class HsSubpartitionDataBuffer {
+        // this field only accessed by task main thread.
+        private final Queue<BufferBuilder> unfinishedBuffers = new LinkedList<>();
+
+        // this field only accessed by task main thread.
+        private int finishedBufferIndex;
+
+        /**
+         * Lock for subpartition's own field. All buffers status belong to this subpartition guarded
+         * by this lock.
+         */
+        private final Object subpartitionLock = new Object();
+
+        // TODO maybe we should rename this to allBuffers or unReleasedBuffers?
+        @GuardedBy("subpartitionLock")
+        private final Deque<BufferContext> finishedBuffers = new LinkedList<>();
+
+        @GuardedBy("subpartitionLock")
+        private final Deque<BufferContext> consumeBuffers = new LinkedList<>();
+
+        @GuardedBy("subpartitionLock")
+        private final Map<Integer, BufferContext> bufferIndexToContexts;
+
+        private final int targetChannel;
+
+        HsSubpartitionDataBuffer(int targetChannel) {
+            this.targetChannel = targetChannel;
+            this.bufferIndexToContexts = new HashMap<>();
+        }
+
+        // ------------------------------------------------------------------------
+        //  Called by Consumer
+        // ------------------------------------------------------------------------
+
+        public Buffer.DataType getFirstDataType(int expectedBufferIndex) {
+            synchronized (subpartitionLock) {
+                while (!consumeBuffers.isEmpty() && consumeBuffers.peekFirst().isReleased) {
+                    consumeBuffers.pollFirst();
+                }
+                if (consumeBuffers.isEmpty()
+                        || consumeBuffers.peekFirst().bufferIndexAndChannel.getBufferIndex()
+                                != expectedBufferIndex) {
+                    return Buffer.DataType.NONE;
+                }
+                BufferContext bufferContext = checkNotNull(consumeBuffers.peekFirst());
+                return bufferContext.buffer.getDataType();
+            }
+        }
+
+        public BufferAndNextDataType pollBuffer(int expectedBufferIndex) {
+            BufferContext bufferContext;
+            Buffer.DataType nextDataType;
+            try {
+                lock.readLock().lock();
```

Comment:
We should only pass `readLock` into the subpartition, to make sure it doesn't accidentally access the write lock.

Reply:
done.